### PR TITLE
Feature/component abstract classes

### PR DESCRIPTION
**Pull request by mmorris44**
## What?
Create abstract classes for components which we will have variants of in our systems (e.g. feed-forward vs recurrent)
## Why?
Abstract classes define templates for components and enforce using consistent names, since the `name()` method comes from the base class
## How?
Defined abstract components for `executor observe`, `executor selection action`, `Network`, `TrainerStep`, `TrainerDataset`, `parameter server`
## Extra
Tested to ensure changes do not break MAPPO system. ~~Wait for [PR #508](https://github.com/instadeepai/Mava/pull/508) to be merged before reviewing and merging this one~~ -> merged

### CODE REVIEW

File: mava/components/jax/building/datasets.py [hunk 1]
```diff
@@ -26,6 +27,38 @@
 Transform = Callable[[reverb.ReplaySample], reverb.ReplaySample]
 
 
+class TrainerDataset(Component):
```
Comment (RuanJohn) [inline mava/components/jax/building/datasets.py:30]: Very minor, but isn't it a bit more intuitive to call this `Dataset `instead?

File: mava/components/jax/executing/action_selection.py [hunk 2]
```diff
@@ -34,14 +32,29 @@ class ExecutorSelectActionProcessConfig:
 
 class ExecutorSelectAction(Component):
     @abc.abstractmethod
-    def __init__(self, config: Any) -> None:
+    def __init__(
+        self,
+        config: ExecutorSelectActionProcessConfig = ExecutorSelectActionProcessConfig(),
```
Comment (RuanJohn) [inline mava/components/jax/executing/action_selection.py:37]: Also minor, but isn't it more intuitive to have it called `ExecutorSelectActionConfig` to match the component name?

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Thank you @mmorris44! Looks good, just a few very minor naming comments.

File: mava/components/jax/building/datasets.py (hunk 1 above)
Comment (mmorris44) [reply mava/components/jax/building/datasets.py:30]: Agreed, changed

File: mava/components/jax/executing/action_selection.py (hunk 2 above)
Comment (mmorris44) [reply mava/components/jax/executing/action_selection.py:37]: Also agreed, changed

**Review by RuanJohn — DISMISSED** [review]

File: mava/components/jax/building/datasets.py (hunk 1 above)
Comment (mmorris44) [reply mava/components/jax/building/datasets.py:30]: Actually, switched it back to `TrainerDataset` to match the hook name. Sorry, that's why I named it that originally

Comment (RuanJohn) [reply mava/components/jax/building/datasets.py:30]: I see, that is fine by me in that case.

**Review by RuanJohn — APPROVED** [review]

**Pull request merged**